command/agent/agent.go

```diff
@@ -1002,9 +1006,15 @@ func (a *Agent) setupClient() error {
 	// running consul-template functions that utilize the Nomad API. We lazy
 	// load this into the client config, therefore this needs to happen before
 	// we call NewClient.
+	//TODO migrate to APIListenerRegistrar
 	a.builtinListener, a.builtinDialer = bufconndialer.New()
 	conf.TemplateDialer = a.builtinDialer
 
+	// Initialize builtin API server here for use in the client, but it won't
+	// accept connections until the HTTP servers are created.
+	a.builtinServer = newBuiltinAPI()
+	conf.APIListenerRegistrar = a.builtinServer
```

Comment:
Is this necessary? The implemenation of `builtinAPI` is clever and unfamiliar, which is usually a bad thing.

Reply:
**tl;dr I would _love_ a better way of doing this.**

> Is this necessary?

Kind of? Basically giving the Client access to the Agent HTTP server creates a circular dependency that needs solving somehow:

1. [NewAgent creates Client](https://github.com/hashicorp/nomad/blob/v1.4.3/command/agent/agent.go#L148-L150)
2. [agent.Command creates HTTP servers](https://github.com/hashicorp/nomad/blob/v1.4.3/command/agent/command.go#L538-L545) which accept the Agent as a dependency

That worked out fine when the Client was oblivious to the existence of the HTTP API.

When we needed access to the HTTP API for `template` support we used [`bufconn` to create a Dialer connected to a Listener](https://github.com/hashicorp/nomad/blob/v1.4.3/command/agent/agent.go#L976-L981), returned the Listener to Agent, and [Agent would slap the HTTP API on it](https://github.com/hashicorp/nomad/blob/v1.4.3/command/agent/http.go#L173-L186). `bufconn` [blocks Dials until the Listener accepts](https://github.com/grpc/grpc-go/blob/v1.52.3/test/bufconn/bufconn.go#L93-L106) them.

Unix Domain Sockets (UDS) are _listeners_ themselves though, so we couldn't just reuse the Dialer (unless we wanted to create 1 Accept goroutine per Task and 2 proxy goroutines for every UDS connection to proxy reads and writes from the UDS to the `bufconn`).

So I more or less copied `bufconn` except instead of blocking a Dial attempt on its corresponding Accept, I block a `Register(listener)` attempt on the HTTP Server existing.

> schmichael why don't you just swap the order we start the api and client?

I think that could work. The `Agent.{Client,Server}` getters for the API to access the client and server would need locks and setters since they would be nil until NewAgent returned them. The dependencies that the API has on Agent itself are harder to untangle: `Agent.RPC(` being one of the harder bits. Do we do a `bufconn`/`builtinServer` style `block RPCs from the HTTP API until the Agent is created`? We don't really gain much from swapping the order if we don't also detangle the circular dependencies. So it gets very complicated especially because both Client and Server have their own RPC implementations and methods of routing requests around and the HTTP API just grabs one and uses it.